algoliasearch-common/src/main/java/com/algolia/search/SyncIndexCRUD.java

```diff
@@ -124,4 +127,111 @@ default Task copyTo(@Nonnull String dstIndexName, @Nonnull List<String> scope)
       throws AlgoliaException {
     return copyTo(dstIndexName, scope, new RequestOptions());
   }
+
+  /**
+   * Rebuild the entirety of your data atomically, to ensure that searches performed on the index
+   * during the rebuild will not be interrupted.
+   *
+   * @param indexContent the content of the index to reindex
+   * @throws AlgoliaException AlgoliaException
+   */
+  default List<Long> reIndex(@Nonnull IndexContent<T> indexContent) throws AlgoliaException {
+    return reIndex(indexContent, false, new RequestOptions());
+  }
+
+  /**
+   * Rebuild the entirety of your data atomically, to ensure that searches performed on the index
+   * during the rebuild will not be interrupted.
+   *
+   * @param indexContent the content of the index to reindex
+   * @param safe runs the reindex safely
+   * @throws AlgoliaException AlgoliaException
+   */
+  default List<Long> reIndex(@Nonnull IndexContent<T> indexContent,boolean safe) throws AlgoliaException {
+    return reIndex(indexContent, safe, new RequestOptions());
+  }
+
+  /**
+   * Rebuild the entirety of your data atomically, to ensure that searches performed on the index
+   * during the rebuild will not be interrupted.
+   *
+   * @param indexContent the content of the index to reindex
+   * @param safe runs the reindex safely
+   * @param requestOptions request options
+   * @throws AlgoliaException Algolia Exception
+   */
+  default List<Long> reIndex(
+      @Nonnull IndexContent<T> indexContent, boolean safe, @Nonnull RequestOptions requestOptions)
+      throws AlgoliaException {
+
+    // 1. Init temps Index
+    Index<T> tmpIndex =
+        this.getApiClient().initIndex(getName() + "_tmp_" + UUID.randomUUID().toString(), indexContent.getObjectClass());
+
+    // 2. Copy the settings, synonyms and rules (but not the records)
+    List<Long> taskIds = new ArrayList<>();
+    List<String> scopes = indexContent.getScopes();
+
+    if (scopes.size() > 0) {
+      Task copyIndexTask = copyTo(tmpIndex.getName(), scopes, requestOptions);
+      taskIds.add(copyIndexTask.getTaskID());
+      if (safe) {
+        tmpIndex.waitTask(copyIndexTask.getTaskID());
+      }
+    }
+
+    if (indexContent.getRules() != null) {
+      Task task = tmpIndex.batchRules(indexContent.getRules(), requestOptions);
+      taskIds.add(task.getTaskID());
+    }
+
+    if (indexContent.getSettings() != null) {
+      Task task = tmpIndex.setSettings(indexContent.getSettings(), requestOptions);
+      taskIds.add(task.getTaskID());
+    }
+
+    if (indexContent.getSynonyms() != null) {
+      Task task = tmpIndex.batchSynonyms(indexContent.getSynonyms(), requestOptions);
+      taskIds.add(task.getTaskID());
+    }
+
+    // 3. Fetch your data with the iterator and push it to the temporary index
+    if (indexContent.getObjects() != null) {
+      ArrayList<T> records = new ArrayList<>();
+
+      for (T object : indexContent.getObjects()) {
+
+        if (records.size() == 1000) {
```

Comment:
Performance issue ( at least on PHP ): You may want to increase a counter instead of calculating the size of the records every object.